File: backend/helpers/scraper.js
const puppeteer = require('puppeteer');
const cheerio = require('cheerio');
const OpenAI = require('openai');

const client = new OpenAI({
    organization: process.env.OPENAI_ORG_ID,
    apiKey: process.env.OPENAI_API_KEY,
});

const scrapeJobPosting = async (jobUrl) => {
    try {
        // Launch a Puppeteer browser instance
        const browser = await puppeteer.launch({ 
            headless: true,
            executablePath: process.env.PUPPETEER_EXECUTABLE_PATH || '/usr/bin/chromium-browser', // Use the system-installed Chromium
            args: ['--no-sandbox', '--disable-setuid-sandbox'] // Disable the sandbox for root execution
        });
        const page = await browser.newPage();
        
        // Go to the job URL
        await page.goto(jobUrl, { waitUntil: 'networkidle0' }); // Wait for the page to fully load

        // Get the full page content
        const data = await page.content();

        // Load the HTML into Cheerio for parsing
        const $ = cheerio.load(data);

        let jobHtml = $('body').html();

        if (jobUrl.includes('myworkdayjobs')) {
            jobHtml = $('.css-141ntgo').html();
        }

        // Close the browser
        await browser.close();

        // Prepare the prompt for ChatGPT API to extract the needed info
        const prompt = `
        Extract the following details from the HTML of a job posting:
        - Job Title
        - Company Name
        - Location
        - Duration of the Position (length)
        - Job Posting URL

        Here is the HTML:
        ${jobHtml}
        `;

        const response = await client.chat.completions.create({
            model: "gpt-3.5-turbo-0125",
            messages: [
                { role: "system", content: "You are a helpful assistant that extracts job details from HTML." },
                { role: "user", content: prompt },
            ],
            max_tokens: 500,
        });

        // Extract and return the result
        const parsedResponse = response.choices[0].message.content;

        // Use regular expressions or string manipulation to extract the values
        const titleMatch = parsedResponse.match(/Job Title:\s*(.*)/);
        const companyMatch = parsedResponse.match(/Company Name:\s*(.*)/);
        const locationMatch = parsedResponse.match(/Location:\s*(.*)/);
        const lengthMatch = parsedResponse.match(/Duration of the Position:\s*(.*)/);
        const postingMatch = parsedResponse.match(/Job Description:\s*([\s\S]*)(?=Job Posting URL)/);  // Capture multi-line job description
        const url = jobUrl;  // Use the provided URL

        // Extract values with fallback in case any field is missing
        const title = titleMatch ? titleMatch[1].trim() : 'Not provided';
        const company = companyMatch ? companyMatch[1].trim() : 'Not provided';
        const location = locationMatch ? locationMatch[1].trim() : 'Not provided';
        const length = lengthMatch ? lengthMatch[1].trim() : 'Not provided';
        const posting = postingMatch ? postingMatch[1].trim() : 'Not provided';

        // Return the result in the required format
        return { title, company, location, length, posting, url };

    } catch (error) {
        console.error('Error scraping job posting:', error);
        return null;
    }
};

module.exports = scrapeJobPosting;

File: backend/app/helpers/scraper.py
import os
import re
from typing import Optional

import requests
from bs4 import BeautifulSoup
from openai import OpenAI
from dotenv import load_dotenv

load_dotenv()

client = OpenAI(
    organization=os.getenv("OPENAI_ORG_ID"),
    api_key=os.getenv("OPENAI_API_KEY"),
)


def scrape_job_posting(job_url: str) -> Optional[dict]:
    try:
        response = requests.get(
            job_url,
            timeout=30,
            headers={
                "User-Agent": (
                    "Mozilla/5.0 (X11; Linux x86_64) "
                    "AppleWebKit/537.36 (KHTML, like Gecko) "
                    "Chrome/124.0.0.0 Safari/537.36"
                )
            },
        )
        response.raise_for_status()

        soup = BeautifulSoup(response.text, "html.parser")

        body = soup.body
        job_html = body.decode_contents() if body else response.text

        if "myworkdayjobs" in job_url:
            workday_section = soup.select_one(".css-141ntgo")
            if workday_section:
                job_html = str(workday_section)

        prompt = f"""
Extract the following details from the HTML of a job posting:
- Job Title
- Company Name
- Location
- Duration of the Position (length)
- Job Description
- Job Posting URL

Here is the HTML:
{job_html}
"""

        ai_response = client.chat.completions.create(
            model="gpt-4o",
            messages=[
                {
                    "role": "system",
                    "content": "You are a helpful assistant that extracts job details from HTML.",
                },
                {"role": "user", "content": prompt},
            ],
            max_tokens=700,
        )

        parsed_response = ai_response.choices[0].message.content or ""

        title_match = re.search(r"Job Title:\s*(.*)", parsed_response)
        company_match = re.search(r"Company Name:\s*(.*)", parsed_response)
        location_match = re.search(r"Location:\s*(.*)", parsed_response)
        length_match = re.search(r"Duration of the Position.*?:\s*(.*)", parsed_response)
        posting_match = re.search(
            r"Job Description:\s*([\s\S]*?)(?=Job Posting URL:|$)",
            parsed_response,
        )

        title = title_match.group(1).strip() if title_match else "Not provided"
        company = company_match.group(1).strip() if company_match else "Not provided"
        location = location_match.group(1).strip() if location_match else "Not provided"
        length = length_match.group(1).strip() if length_match else "Not provided"
        posting = posting_match.group(1).strip() if posting_match else "Not provided"
        url = job_url

        print("Scraped job posting:", {
            "title": title,
            "company": company,
            "location": location,
            "length": length,
            "posting": posting,
            "url": url,
        })

        return {
            "title": title,
            "company": company,
            "location": location,
            "length": length,
            "posting": posting,
            "url": url,
        }

    except Exception as error:
        print("Error scraping job posting:", error)
        return None